### PR TITLE
2226: Make it possible to avoid "forward backports"

### PR DESCRIPTION
**Pull request by edvbld**
Hi all,

please review this re-implementation of avoiding so called "forward backports", this time around in the `IssueNotifier`, meaning that Skara will avoid to create a "forward backport" _after_ a pull request has been integrated. Solving the problem in the `IssueNotifier` turned out to be a much more natural fit, so thanks @zhaosongzs and @erikj79 for the discussion in #1630.

The new implementation will avoid creating a "foward backport" if _no_ suitable backport (or "forward backport") is found. That is, if JBS already contains a backport (or "forward backport") that can be used, then that backport (or "forward backport") will be used. This makes sense to me because in this scenario the user has (most likely) set up the issues in JBS for this to happen. But when Skara decides to create a _new_ backport, then (if `"avoidforwardports"` is `true` in the configuration) the code will never create "forward backport", instead opting to create a backport with the primary issue's `fixVersions` and setting `fixVersions` of the primary issue to the value found in the repository's `.jcheck/conf`.

I also added a number of new unit tests. Note that this pull request depends on #1636 in order for the tests to pass, so the tests will fail until #1636 is integrated.

Thanks,
Erik

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace

### Issue
 * [SKARA-2226](https://bugs.openjdk.org/browse/SKARA-2226): Make it possible to avoid "forward backports" (**Enhancement** - P4)

### Reviewers
 * [Zhao Song](https://openjdk.org/census#zsong) (@zhaosongzs - **Reviewer**)
 * [Erik Joelsson](https://openjdk.org/census#erikj) (@erikj79 - **Reviewer**) ⚠️ Review applies to [26b3bf4c](https://git.openjdk.org/skara/pull/1637/files/26b3bf4c7a5eb4302712da50fe518e1a552ddcc1)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/skara.git pull/1637/head:pull/1637` \
`$ git checkout pull/1637`

Update a local copy of the PR: \
`$ git checkout pull/1637` \
`$ git pull https://git.openjdk.org/skara.git pull/1637/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1637`

View PR using the GUI difftool: \
`$ git pr show -t 1637`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/skara/pull/1637.diff">https://git.openjdk.org/skara/pull/1637.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/skara/pull/1637#issuecomment-2051329703)

### CODE REVIEW

File: bots/notify/src/test/java/org/openjdk/skara/bots/notify/issue/IssueNotifierTests.java [hunk 1]
```diff
@@ -2238,4 +2238,264 @@ void testFailedIssue(TestInfo testInfo) throws IOException {
...
+            // Create an issue and commit a fix
+            var issue = issueProject.createIssue("This is an issue", List.of("Indeed"),
+                    Map.of("issuetype", JSON.of("Enhancement"),
+                            SUBCOMPONENT, JSON.of("java.io"),
+                            RESOLVED_IN_BUILD, JSON.of("b07")
+                    ));
+            issue.setProperty("fixVersions", JSON.array().add("22"));
+            issue.setProperty("priority", JSON.of("1"));
+            issue.addLabel("test");
+            issue.addLabel("temporary");
```
Comment (zhaosongzs) [inline bots/notify/src/test/java/org/openjdk/skara/bots/notify/issue/IssueNotifierTests.java:2330]: In this case, maybe it's better to set the main issue as RESOLVED

File: bots/notify/src/test/java/org/openjdk/skara/bots/notify/issue/IssueNotifierTests.java [hunk 2]
```diff
@@ -2238,4 +2238,264 @@ void testFailedIssue(TestInfo testInfo) throws IOException {
...
+
+            // Create an explicit "forwardport"
```
Comment (zhaosongzs) [inline bots/notify/src/test/java/org/openjdk/skara/bots/notify/issue/IssueNotifierTests.java:2462]: Should be `Create an explicit "backport"`

**Review by zhaosongzs — COMMENTED** [review]
Overall this patch looks good!
I believe the issueNotifier would work well even when the bot gets interrupted.
Will approve it after SKARA-2226 gets integrated and all the GitHub actions pass.

**Review by erikj79 — APPROVED** [review]

File: bots/notify/src/test/java/org/openjdk/skara/bots/notify/issue/IssueNotifierTests.java (hunk 2 above)
Comment (edvbld) [reply bots/notify/src/test/java/org/openjdk/skara/bots/notify/issue/IssueNotifierTests.java:2462]: Thanks, will fix 👍

File: bots/notify/src/test/java/org/openjdk/skara/bots/notify/issue/IssueNotifierTests.java (hunk 1 above)
Comment (zhaosongzs) [reply bots/notify/src/test/java/org/openjdk/skara/bots/notify/issue/IssueNotifierTests.java:2330]: Ah, got it. NVM

**Review by zhaosongzs — APPROVED** [review]

**Comment by edvbld** [comment]
Thanks @zhaosongzs and @erikj79 for reviewing!

**Pull request closed without merging**